ElectricCar.cpp

```diff
@@ -1,16 +1,30 @@
 #include "ElectricCar.hpp"
 #include <iostream>
+#include <stdexcept>
 
 ElectricCar::ElectricCar(ElectricEngine* engine)
     : engine_(engine)
 {
     std::cout << __FUNCTION__ << std::endl;
 }
 
-ElectricCar::~ElectricCar()       { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnLeft()      { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::turnRight()     { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::brake()         { std::cout << __FUNCTION__ << std::endl; }
-void ElectricCar::accelerate(int) { std::cout << __FUNCTION__ << std::endl; }
+ElectricCar::~ElectricCar() {
+    delete engine_;
+    std::cout << __FUNCTION__ << std::endl;
+}
+void ElectricCar::restore()       { charge(); }
 void ElectricCar::charge()        { std::cout << __FUNCTION__ << std::endl; }
+void ElectricCar::setEngine(int power, int batteryCapacity) {
+    if (Car::speed_ == 0) {
+        engine_ = new ElectricEngine(power, batteryCapacity);
+        std::cout << "Engine changed" << std::endl;
+    }
+    else {
+        throw std::logic_error("Can't change engine while driving (please verify with Youtube)");
+    }
+}
```

Comment:
nice solution for a problem with providing different kind of engine. However, what if the design forbids it? Eg. we have an EngineFactory class that should produce an engine and we are obliged to get a pointer? It's only hypothetical question ;)